plugins/protocols/session-lock.cpp

```diff
@@ -0,0 +1,434 @@
+        }
+
+        void lock_all()
+        {
+            for (const auto& [output, output_state] : output_states)
+            {
+                lock_output(output, output_state);
+            }
+
+            wlr_session_lock_v1_send_locked(lock);
+            set_state(LOCKED);
+        }
+
+        void remove_backup_surfaces()
+        {
+            for (const auto& [output, output_state] : output_states)
+            {
+                if (output_state->backup_surface)
+                {
+                    wf::scene::remove_child(output_state->backup_surface);
+                    output_state->backup_surface.reset();
+                }
+            }
+        }
+
+        void unlock_all()
+        {
+            remove_backup_surfaces();
+            for (const auto& [output, output_state] : output_states)
+            {
+                output->set_inhibited(false);
+            }
+
+            set_state(UNLOCKED);
+            LOGD("unlock");
+        }
+
+        void set_state(lock_state new_state)
+        {
+            state = new_state;
+            plugin->notify_lock_state(state);
+        }
+
+        wf_session_lock_plugin *plugin;
+        wlr_session_lock_v1 *lock;
+        wf::wl_timer<false> lock_timer;
+        std::map<wf::output_t*, std::shared_ptr<output_state>> output_states;
+
+        wf::wl_listener_wrapper new_surface;
+        wf::wl_listener_wrapper unlock;
+        wf::wl_listener_wrapper destroy;
+
+        wf::signal::connection_t<wf::output_added_signal> output_added;
+        wf::signal::connection_t<wf::output_removed_signal> output_removed;
+    };
+
+  public:
+    void init() override
+    {
+        auto display = wf::get_core().display;
+        manager = wlr_session_lock_manager_v1_create(display);
+
+        new_lock.set_callback([this] (void *data)
+        {
+            wlr_session_lock_v1 *wlr_lock = (wlr_session_lock_v1*)data;
+
+            if (cur_lock.get() == nullptr)
+            {
+                cur_lock.reset(new wayfire_session_lock(this, wlr_lock));
+                LOGD("new_lock");
+            } else
+            {
+                LOGE("new_lock: already locked");
+                wlr_session_lock_v1_destroy(wlr_lock);
+            }
+        });
+        new_lock.connect(&manager->events.new_lock);
+
+        destroy.set_callback([this] (void *data)
+        {
+            LOGI("session_lock_manager destroyed");
+        });
+        destroy.connect(&manager->events.destroy);
+    }
+
+    void notify_lock_state(lock_state state)
+    {
+        switch (state)
+        {
+          case UNLOCKED:
+            prev_lock.reset();
+            break;
+
+          case DESTROYED:
+            cur_lock.reset();
+            break;
+
+          case ZOMBIE:
+            prev_lock = std::move(cur_lock);
+            break;
+
+          default:
+            break;
+        }
+    }
+
+    void fini() override
+    {
+        // TODO: unlock everything?
```

Comment:
I don't think we actually need to do anything. Plugins that cannot be unloaded like this one will be destroyed only at compositor shutdown. At that point the wayland display is being destroyed, so wlroots will auto-destroy the protocol and it will emit destruction signals for the various lock surfaces, etc. (I hope it does so at least).